src/backend/distributed/planner/shard_pruning.c

```diff
@@ -825,6 +852,45 @@ AddPartitionKeyRestrictionToInstance(ClauseWalkerContext *context, OpExpr *opCla
 }
 
 
+/*
+ * Sometimes PostgreSQL chooses to try to wrap our Var in a coercion rather
+ * than the Const; to deal with this, we strip the coercions from both and
+ * manually coerce the Const into the type of our partition column. It is
+ * conceivable that in some instances, this may not be possible; in those cases
+ * we will simply fail to prune partitions based on this clause.
+ */
+static Const *
+TransformPartitionRestrictionValue(Var *partitionColumn, Node *restrictionValue)
+{
+	Node *transformedValue = coerce_to_target_type(NULL, restrictionValue,
+												   exprType(restrictionValue),
+												   partitionColumn->vartype,
+												   partitionColumn->vartypmod,
+												   COERCION_ASSIGNMENT,
+												   COERCE_IMPLICIT_CAST, -1);
+
+	/* if NULL, no implicit coercion is possible between the types */
+	if (transformedValue == NULL)
+	{
+		return NULL;
+	}
+
+	/* if still not a constant, evaluate coercion */
+	if (!IsA(transformedValue, Const))
+	{
+		transformedValue = (Node *) expression_planner((Expr *) transformedValue);
```

Comment:
Nice, I haven't seen this function before! I'll probably update [here](https://github.com/citusdata/citus/blob/master/src/backend/distributed/planner/fast_path_router_planner.c#L85-L92) to use it instead of `eval_const_expressions`.

Reply:
Yeah it seemed like a _slightly_ higher-level abstraction.